### PR TITLE
ENH: galaxy-quenching model

### PR DESCRIPTION
**Pull request by Lucia-Fonseca**
## Description
This PR adds a feature where we can obtain the Schechter parameters for active galaxies (centrals and satellites) and passive galaxies (mass- and satellite-quenched) based on equation (15) in de la Bella et al. 2021. Merging this PR coses #513 .

## Checklist
- [ ] Follow the [Contributor Guidelines](https://github.com/skypyproject/skypy/blob/main/CONTRIBUTING.rst)
- [ ] Write unit tests
- [ ] Write documentation strings
- [ ] Assign someone from your working team to review this pull request
- [ ] Assign someone from the infrastructure team to review this pull request

### CODE REVIEW

File: skypy/galaxies/stellar_mass.py [hunk 1]
```diff
@@ -74,3 +75,58 @@ def schechter_smf_mass(redshift, alpha, m_star, m_min, m_max, size=None,
     m = schechter(alpha, x_min, x_max, resolution, size=size, scale=m_star)
 
     return m
+
+
+def schechter_smf_parameters(active_parameters, fsatellite, fenvironment):
```
Comment (philipp128) [inline skypy/galaxies/stellar_mass.py:80]: I know a bit picky but maybe change the order of `fsatellite` and `fenvironment` to match the order in equation (15) of your paper. You do not need to change this but just an idea to be consistent.

File: skypy/galaxies/stellar_mass.py [hunk 2]
```diff
@@ -74,3 +75,58 @@ def schechter_smf_mass(redshift, alpha, m_star, m_min, m_max, size=None,
...
+
+
+    Returns
+    -------
+    parameters: dic
+        It returns a  dictionary with the parameters of the
+        Schechter mass function. The dictionary keywords:
+        `centrals`, `satellites`, `mass_quenched` and
+        `satellite_quenched`. The values correspond to a
+        tuple :math:`(\phi, \alpha, m_{*})`.
```
Comment (philipp128) [inline skypy/galaxies/stellar_mass.py:112]: I am confused. Is it meant to return `phi` or `phi_star`? Looking at equation (15) I assumed we want `phi_star`.

File: skypy/galaxies/stellar_mass.py [hunk 3]
```diff
@@ -74,3 +75,58 @@ def schechter_smf_mass(redshift, alpha, m_star, m_min, m_max, size=None,
...
+    This function returns the Schechter mass function parameters
+    for active galaxies (centrals and satellites)
+    and passive galaxies (mass- and satellite-quenched)
+    based on equation (15) in de la Bella et al. 2021 [1]_.
+
+    Parameters
+    ----------
+    active_parameters: tuple
+        Schechter mass function parameters for the entire active
+        sample of galaxies: :math:`(\phi_b, \alpha_b, m_{*})`.
```
Comment (philipp128) [inline skypy/galaxies/stellar_mass.py:91]: I suggest you call the parameter phi_{*,b}. I know it does not make a real difference but then it is consistent equation (15) in the paper.
```suggestion
        sample of galaxies: :math:`(\phi_{*,b}, \alpha_b, m_{*})`.
```

File: skypy/galaxies/stellar_mass.py [hunk 4]
```diff
@@ -74,3 +75,58 @@ def schechter_smf_mass(redshift, alpha, m_star, m_min, m_max, size=None,
...
+        Note: in [1], :math:`\alpha_b` and :math:`m_{*}` are
+        constants. The elements of the tuple should be scalars.
+        Exception: :math:`\phi_b` can be array only when
```
Comment (philipp128) [inline skypy/galaxies/stellar_mass.py:94]: Here the same as above.

File: skypy/galaxies/stellar_mass.py [hunk 5]
```diff
@@ -74,3 +75,58 @@ def schechter_smf_mass(redshift, alpha, m_star, m_min, m_max, size=None,
...
+        `satellite_quenched`. The values correspond to a
+        tuple :math:`(\phi, \alpha, m_{*})`.
+
+    References
+    ----------
+    .. [1] de la Bella et al. 2021, Quenching and Galaxy Demographics,
+           arXiv 2112.11110.
+
+    '''
+    phi, alpha, mstar = active_parameters
```
Comment (philipp128) [inline skypy/galaxies/stellar_mass.py:120]:
```suggestion
    phi_star, alpha, mstar = active_parameters
```

File: skypy/galaxies/stellar_mass.py [hunk 6]
```diff
@@ -74,3 +75,58 @@ def schechter_smf_mass(redshift, alpha, m_star, m_min, m_max, size=None,
...
+
+    sum_phics = (1 - fsatellite) * (1 - np.log(1 - fsatellite))
+    phic = (1 - fsatellite) * phi / sum_phics
+    phis = phic * np.log(1 / (1 - fsatellite))
```
Comment (philipp128) [inline skypy/galaxies/stellar_mass.py:124]:
```suggestion
    phic_star = (1 - fsatellite) * phi / sum_phics
    phis_star = phic * np.log(1 / (1 - fsatellite))
```

File: skypy/galaxies/stellar_mass.py [hunk 7]
```diff
@@ -74,3 +75,58 @@ def schechter_smf_mass(redshift, alpha, m_star, m_min, m_max, size=None,
...
+
+    centrals = (phic, alpha, mstar)
+    satellites = (phis, alpha, mstar)
+    mass_quenched = (phic + phis, alpha + 1, mstar)
+    satellite_quenched = (- np.log(1 - fenvironment) * phis, alpha, mstar)
```
Comment (philipp128) [inline skypy/galaxies/stellar_mass.py:129]: Changes according to above comments

File: skypy/galaxies/tests/test_stellar_mass.py [hunk 8]
```diff
@@ -60,3 +60,61 @@ def calc_cdf(m):
                                              size=1000, resolution=100)
     p_value = scipy.stats.kstest(sample, calc_cdf)[1]
     assert p_value >= 0.01
+
+
+def test_schechter_smf_parameters():
+    from skypy.galaxies.stellar_mass import schechter_smf_parameters
+    # Check scalar inputs
+    blue = (10**-2.423, 10**10.60, -1.21)
+    fsat_scalar, frho = 0.4, 0.2
```
Comment (philipp128) [inline skypy/galaxies/tests/test_stellar_mass.py:69]: Maybe be consistent with naming convention in method.

```suggestion
    fsatellite_scalar, fenvironment = 0.4, 0.2
```

**Review by philipp128 — CHANGES_REQUESTED** [review]
For the moment I have mainly comments on naming convention. I found it first hard to figure out what is calculated especially cause of the names in the paper. At the end it does not make a big difference but it was confusing to me.

**Review by rrjbca — COMMENTED** [review]
I would suggest refactoring as follows:
- mstar is not used to calculate any of the phi or alpha values and is returned unchanged in all four populations. I would remove mstar from the arguments and return values.
- Similarly, alpha is not used to calculate any of the phi values and is only modified from the input value for one population. That function is trivial (1 + alpha), does not depend on phi, fsatellite or fenvironment, and is a result of a very specific modelling assumption (quenching rate proportional to mass). I would at least move the alpha calculation to a separate function, and probably suggest removing it entirely.
- This leaves just the calculation of the phi values. The active_parameters tuple can be replaced with just phi since there is no longer any dependence on mstar or alpha. I would also split into four functions: phi_centrals, phi_satellites, phi_mass_quenched and phi_satellite_quenched. This makes it much clearer from the name what each function returns and saves the user having to check the dictionary keys in the documentation when calling it. Individual functions can also be used more easily in situations where the user doesn't want to model all four populations.

File: skypy/galaxies/stellar_mass.py [hunk 9]
```diff
@@ -74,3 +78,131 @@ def schechter_smf_mass(redshift, alpha, m_star, m_min, m_max, size=None,
...
+        Amplitude of the Schechter mass function.
+
+    References
+    ----------
+    .. [1] de la Bella et al. 2021, Quenching and Galaxy Demographics,
+           arXiv 2112.11110.
+
+    '''
+
+    return phi_centrals + phi_satellites
```
Comment (rrjbca) [inline skypy/galaxies/stellar_mass.py:178]: Do we definitely need such a simple function?

File: skypy/galaxies/stellar_mass.py [hunk 10]
```diff
@@ -74,3 +78,131 @@ def schechter_smf_mass(redshift, alpha, m_star, m_min, m_max, size=None,
...
+        Schechter mass function amplitude of the satellite
+        active galaxies.
+
+    fenvironment: float
+        Fraction of satellite-quenched galaxies between 0 and 1.
+
+    Returns
+    -------
+    amplitude: float, (nt, nm) array_like
+        Amplitude of the Schechter mass function.
```
Comment (rrjbca) [inline skypy/galaxies/stellar_mass.py:199]:
```suggestion
    phi_satellites: array_like
        Schechter mass function amplitude of the satellite
        active galaxies.

    fenvironment: array_like
        Fraction of satellite-quenched galaxies between 0 and 1.

    Returns
    -------
    amplitude: array_like
        Amplitude of the Schechter mass function.
```
Surely `fenvironment` could also be an array here?

File: skypy/galaxies/stellar_mass.py [hunk 11]
```diff
@@ -74,3 +78,131 @@ def schechter_smf_mass(redshift, alpha, m_star, m_min, m_max, size=None,
...
+           arXiv 2112.11110.
+
+    '''
+
+    if np.ndim(phi_blue_total) == 1 and np.ndim(fsatellite) == 1:
+        phi_blue_total = phi_blue_total[:, np.newaxis]
+
+    sum_phics = (1 - fsatellite) * (1 - np.log(1 - fsatellite))
+
+    return (1 - fsatellite) * phi_blue_total / sum_phics
```
Comment (rrjbca) [inline skypy/galaxies/stellar_mass.py:116]:
```suggestion
    return phi_blue_total / (1 - np.log(1 - fsatellite))
```

File: skypy/galaxies/stellar_mass.py [hunk 12]
```diff
@@ -74,3 +78,131 @@ def schechter_smf_mass(redshift, alpha, m_star, m_min, m_max, size=None,
...
+        Amplitude of the Schechter mass function.
+
+    References
+    ----------
+    .. [1] de la Bella et al. 2021, Quenching and Galaxy Demographics,
+           arXiv 2112.11110.
+
+    '''
+
+    return - np.log(1 - fenvironment) * phi_satellites
```
Comment (rrjbca) [inline skypy/galaxies/stellar_mass.py:208]: This is the same as `schechter_smf_amplitude_satellites` so perhaps we can get rid of this function?

**Review by rrjbca — CHANGES_REQUESTED** [review]
See comments for some possible further refactoring. The names of the functions seem very long can we think of anything shorter? E.g. perhaps `_amplitude_` -> `_phi_` ?

Regarding your unit tests you do far more tests of the output shapes than is really necessary. For each function one test with scalar inputs checking that it returns a scalar and another test with array inputs to check that the shapes broadcast together correctly. Any more doesn't really add anything.

On the other hand there aren't enough tests that the functions do numerically what you expect. You have one good example where `fsat/frho =0` and I would want to see more like this. E.g. for test_schechter_smf_amplitude_centrals If `fsatellite = 1 - np.exp(-1)` then you expect the function to return `0.5 * phi_blue_total`. I think for each function you should be able to come up with the input "f" values that scale the amplitude by 0, 0.5 and 1 and test them.

File: skypy/galaxies/stellar_mass.py [hunk 13]
```diff
@@ -74,3 +78,131 @@ def schechter_smf_mass(redshift, alpha, m_star, m_min, m_max, size=None,
...
+        sample of galaxies, :math:`(\phi_b, \alpha_b, m_{*})`.
+
+    fsatellite: float, (nm, ) array_like
+        Fraction of active satellite galaxies between 0 and 1.
+        It could be a float or an array, depending on the model you choose.
+
+    Returns
+    -------
+    amplitude: float, (nt, nm) array_like
+        Amplitude of the Schechter mass function.
```
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/galaxies/stellar_mass.py:102]: @rrjbca  It might be the case where phi will have a time and a mass dependence. `phi_blue_total` could be a scalar but it can also be an array of size `nt`. Likewise, `fsatellite` can be a scalar or an array of size `nm`.  If `nt != nm` the product cannot be broadcasted.

File: skypy/galaxies/stellar_mass.py [hunk 14]
```diff
@@ -74,3 +78,131 @@ def schechter_smf_mass(redshift, alpha, m_star, m_min, m_max, size=None,
...
+        active galaxies.
+
+    fsatellite: float, (nm, ) array_like
+        Fraction of active satellite galaxies between 0 and 1.
+        It could be a float or an array, depending on the model you choose.
+
+    Returns
+    -------
+    amplitude: float, (nt, nm) array_like
+        Amplitude of the Schechter mass function.
```
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/galaxies/stellar_mass.py:138]: same as above

File: skypy/galaxies/stellar_mass.py (hunk 9 above)
Comment (Lucia-Fonseca) [reply skypy/galaxies/stellar_mass.py:178]: Not necessarily, but I would add it for the completeness of the model.

File: skypy/galaxies/stellar_mass.py (hunk 10 above)
Comment (Lucia-Fonseca) [reply skypy/galaxies/stellar_mass.py:199]: Could be, `fenvironment` could depend on time, although not mass.

File: skypy/galaxies/stellar_mass.py (hunk 12 above)
Comment (Lucia-Fonseca) [reply skypy/galaxies/stellar_mass.py:208]: is it the same?

Comment (Lucia-Fonseca) [reply skypy/galaxies/stellar_mass.py:208]: it uses `phi_satellites` that one would presumably compute with `schechter_smf_amplitude_satellites` but it has a different prefactor.

File: skypy/galaxies/stellar_mass.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (rrjbca) [reply skypy/galaxies/stellar_mass.py:102]: I don't see any reason not to make the function more flexible and allow it to take any two arrays with broadcastable shapes. If you have two 1D arrays and want the result for every combination the way to broadcast them would be:
```python
>>> x = np.arange(5)
>>> y = np.arange(7)
>>> result = schechter_smf_amplitude_centrals(x, y[:, np.newaxis])
>>> result.shape
(5, 7)
```

File: skypy/galaxies/stellar_mass.py (hunk 12 above)
Comment (rrjbca) [reply skypy/galaxies/stellar_mass.py:208]: `-log(1-f)` is equal to `log(1/(1-f))` so both are functionally identical

**Comment by Lucia-Fonseca** [comment]
@rrjbca do I need to make `hypothesis` a new dependency of skypy to run the tests?
Also I'm a bit confused about the merge conflict.
Thanks for your help.

**Comment by Lucia-Fonseca** [comment]
Hi @itrharrison I can see your last commits but I do not understand why they should be part of this PR?

**Comment by itrharrison** [comment]
@Lucia-Fonseca they are fixes to failing codestyle tests which had appeared because (I think) a newer version of flake8 being more strict about whitespace next to keywords like `assert`. They are indeed nothing to do with this PR but I fixed them so that the tests ran okay.

**Comment by Lucia-Fonseca** [comment]
Yes, I'd wait for the rest of their approvals. 
Nonetheless, I think if a new flake8 version is making some tests or code fail but not this one on this PR, a new independent issue should be open, along with a different PR. Unless I am mistaken, @rrjbca ?

**Comment by itrharrison** [comment]
@Lucia-Fonseca see #572 :-)

**Comment by itrharrison** [comment]
@Lucia-Fonseca I think the only thing missing to fulfill @rrjbca's review requests is adding corner cases for the `schechter_smf_phi_mass_quenched` function. I think these testsdo that by the letter of the law, at least 😝 
```
assert phi0 == 0.5 * schechter_smf_phi_mass_quenched(phi0, phi0)
assert 0.0 == schechter_smf_phi_mass_quenched(phi0, -phi0)
assert phi0 == 1. * schechter_smf_phi_mass_quenched(phi0, 0.0)
```

**Comment by Lucia-Fonseca** [comment]
@itrharrison I just submitted these changes.

> @Lucia-Fonseca I think the only thing missing to fulfill @rrjbca's review requests is adding corner cases for the `schechter_smf_phi_mass_quenched` function. I think these tests do that by the letter of the law, at least 😝
> 
> ```
> assert phi0 == 0.5 * schechter_smf_phi_mass_quenched(phi0, phi0)
> assert 0.0 == schechter_smf_phi_mass_quenched(phi0, -phi0)
> assert phi0 == 1. * schechter_smf_phi_mass_quenched(phi0, 0.0)
> ```
> 
> If you do these I will be happy to dismiss Richard's review if he doesn't respond soon himself 😃
